### PR TITLE
Add decodeEvents function

### PR DESCRIPTION
**Pull request by qperrot**
## Usage related changes

-   Implementation of decodeEvents function, which return an array of objects, containing the data of all emitted events.
-  Check the branch to see the test [`starknet-hardhat-example`] (https://github.com/Shard-Labs/starknet-hardhat-example/pull/68)

## Checklist:

-   [x] Formatted the code
-   [x] No linter errors
-   [x] Tried to avoid introducing linter warnings
-   [x] Performed a self-review of the code
-   [x] Rebased to the last commit of the target branch (or merged it into my branch)
-   [x] Documented the changes
-   [x] Updated the `test` directory
-   [x] Created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/Shard-Labs/starknet-hardhat-example):
    -   [x] Modified `test.sh` to use my example repo branch
    -   [ ] Restored `test.sh` to to use the original branch (after the example repo PR has been merged)
-   [x] All tests are passing

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -700,4 +700,50 @@ export class StarknetContract {
...
+            rawResultIndex++;
+        }
+        const eventNames: string[] = [];
+        rawResultIndex = 0;
+        for (const functionName in this.abi) {
+            const func = this.abi[functionName];
+            if (func.type == "event") {
+                for (const event of events) {
+                    const result = hash.getSelectorFromName(func.name);
+                    if (event.keys[0] == result) {
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply src/types.ts:740]: Add a comment above this comparison that justifies why you're using index 0 (something like `// encoded event name guaranteed to be at index 0`)

File: src/types.ts [hunk 2]
```diff
@@ -700,4 +700,50 @@ export class StarknetContract {
...
+        }
+        return events;
+    }
+
+    /**
+     * decode Events to an array with all the arguments emmitted by the events.
+     * @param events array of unstructured events
+     * @returns array of the arguments emmitted by the events
+     */
+    async decodeEvents(events: starknet.Event[]): Promise<StringMap[]> {
```
Comment (FabijanC) [inline src/types.ts:725]: Let's make it return `Promise<DecodedEvent[]>`, where `DecodedEvent` would hold the event name under `name` and could have a property `data` of type `StringMap`.

**Review by FabijanC — CHANGES_REQUESTED** [review]
In the checklist, you said you modified the `test` directory, but I don't see any changes to it. Your example repo test never got executed because you didn't introduce a test case for it here, in the plugin repo (remember your last PR and how there has to be a check.sh script).

File: src/types.ts [hunk 3]
```diff
@@ -700,4 +728,42 @@ export class StarknetContract {
...
+     * @returns structured object with parameter names.
+     */
+    async decodeEvents(events: starknet.Event[]): Promise<DecodedEvent[]> {
+        const rawEvents: string[] = [];
+        const eventNames: string[] = [];
+        for (const event of events) {
+            const result = event.data.map(BigInt).join(" ");
+            rawEvents.push(result);
+            // encoded event name guaranteed to be at index 0
+            const key = this.eventsSpecifications[event.keys[0]];
```
Comment (FabijanC) [inline src/types.ts:764]: Don't use a generic name as `key`. Call it `eventSpecification`.

Comment (FabijanC) [reply src/types.ts:764]: Also, do a check and throw a HardhatPluginError (for reference, see other places where this is done) if the key doesn't exist in this.eventsSpecifications.

File: src/starknet-types.ts [hunk 4]
```diff
@@ -28,7 +28,14 @@ export interface CairoFunction {
     outputs: Argument[];
 }
 
-export type AbiEntry = CairoFunction | Struct;
+export interface CairoEvent {
+    data: Argument[];
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply src/starknet-types.ts:32]: I don't think you addressed this.

File: account-contract-artifacts/ArgentAccount/0.2.2/ArgentAccount.cairo/ArgentAccount.json [hunk 5]
```diff
@@ -779,11 +779,7 @@
                 "value": "guardian signature invalid"
             }
         ],
-        "builtins": [
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply account-contract-artifacts/ArgentAccount/0.2.2/ArgentAccount.cairo/ArgentAccount.json:782]: I don't think you addressed this properly

File: src/starknet-types.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (qperrot) [reply src/starknet-types.ts:32]: data is an array of object with properties name and type, so I should use `Argument`

File: account-contract-artifacts/ArgentAccount/0.2.2/ArgentAccount.cairo/ArgentAccount.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (FabijanC) [reply account-contract-artifacts/ArgentAccount/0.2.2/ArgentAccount.cairo/ArgentAccount.json:782]: This file (ArgentAccount.json) should be in the same state as it is on the master branch, this PR shouldn't modify it. The link that you provided shows the initial changes that you made, but to correctly respond to my comment, you should link to a commit which undoes this change.

File: src/types.ts (hunk 3 above)
Comment (qperrot) [reply src/types.ts]: Name has been changed and check error is added https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/130/commits/d8ab7d892c576df6b5dd509ea54cd88f8526f065#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410a

**Pull request merged**